org.jacoco.core.test/src-kotlin/org/jacoco/core/test/validation/KotlinGeneratedTest.java

```diff
@@ -0,0 +1,43 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2018 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Fabian Mastenbroek - initial API and implementation
+ *    
+ *******************************************************************************/
+package org.jacoco.core.test.validation;
+
+import static org.junit.Assert.assertNotNull;
+
+import org.jacoco.core.analysis.IClassCoverage;
+import org.jacoco.core.analysis.ICounter;
+import org.jacoco.core.test.validation.targets.KotlinDataClassTarget;
+import org.jacoco.core.test.validation.targets.KotlinDataClassTargetKt;
+import org.junit.Test;
+
+/**
+ * Tests of generated methods for a data class.
+ */
+public class KotlinGeneratedTest extends ValidationTestBase {
+
+	public KotlinGeneratedTest() {
+		super("src-kotlin", KotlinDataClassTargetKt.class);
+	}
+
+	@Test
+	public void testCovered() {
+		final IClassCoverage classCoverage =
+				coverageForClass(KotlinDataClassTarget.CoveredDataClass.class);
```

Comment:
@Godin I would also prefer to stick with validation on source lines only. I add another point:

* I would like to propose to add the expected coverage directly in the source file in future instead of a separate Test class. This makes our integration test suite more maintainable.

Testing method exclusion is more tricky as there can be multiple methods on the same line. So you won't notice if a new method sneaks in on a line where we expect a method to be (think of default ctor and clinit).

Reply:
> Testing method exclusion is more tricky as there can be multiple methods on the same line. So you won't notice if a new method sneaks in on a line where we expect a method to be (think of default ctor and clinit).

Maybe just a slight change to my proposal can cover this case - each method must have at least one **unique** line marker.

> I would like to propose to add the expected coverage directly in the source file in future instead of a separate Test class.

This might be tricky, because in some cases result depends on version of compiler. But indeed let's consider this "in future".

Reply:
If we want to test for existing of methods in the coverage report probably the source file is the wrong place. I would prefer explicit assertions. E.g. `ValidationTestBase` can get a new assert method that takes a list of (class,method) tuples expected in the coverage report for a given source file.